signac/core/collection_api.py

```diff
@@ -0,0 +1,491 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+import os
+import json
+import errno
+import uuid
+from copy import copy
+from contextlib import contextmanager
+from collections.abc import Mapping
+from collections.abc import MutableMapping
+from collections.abc import Sequence
+from collections.abc import MutableSequence
+from abc import abstractmethod
+
+try:
+    import numpy
+    NUMPY = True
+except ImportError:
+    NUMPY = False
+
+try:
+    from collections.abc import Collection
+except ImportError:
+    # Collection does not exist in Python 3.5, only Python 3.6 or newer.
+
+    from collections.abc import Sized, Iterable, Container
+
+    def _check_methods(C, *methods):
+        mro = C.__mro__
+        for method in methods:
+            for B in mro:
+                if method in B.__dict__:
+                    if B.__dict__[method] is None:
+                        return NotImplemented
+                    break
+            else:
+                return NotImplemented
+        return True
+
+    class Collection(Sized, Iterable, Container):  # type: ignore
+        @classmethod
+        def __subclasshook__(cls, C):
+            if cls is Collection:
+                return _check_methods(C,  "__len__", "__iter__", "__contains__")
+            return NotImplemented
+
+
+class SyncedCollection(Collection):
+
+    def __init__(self, parent=None):
+        self._data = None
+        self._suspend_sync_ = 0
+        self._parent = parent
+
+    # TODO add back-end
+    @classmethod
+    def from_base(self, data, filename=None, parent=None):
+        if isinstance(data, Mapping):
+            return JSONDict(filename=filename, data=data, parent=parent)
+        elif isinstance(data, Sequence) and not isinstance(data, str):
+            return JSONList(filename=filename, data=data, parent=parent)
+        elif NUMPY:
+            if isinstance(data, numpy.number):
+                return data.item()
+            elif isinstance(data, numpy.ndarray):
+                return JSONList(filename=filename, data=data.tolist(), parent=parent)
+        return data
+
+    @abstractmethod
+    def to_base(self):
+        pass
+
+    @contextmanager
+    def _suspend_sync(self):
+        self._suspend_sync_ += 1
+        yield
+        self._suspend_sync_ -= 1
+
+    @abstractmethod
+    def _load(self):
+        pass
+
+    @abstractmethod
+    def _sync(self):
+        pass
+
+    @contextmanager
+    def _safe_sync(self):
+        backup = self._data
+        try:
+            yield
+        except BaseException:
+            self._data = backup
+            raise
```

Comment:
The way that this is currently being used is not actually safe. Since `self._data` is a dict, `backup` will just be a reference to that dict, so any modifications will still persist. For now, I think the only easy option is making a copy of `self._data`; this will be a significant performance hit, so we'll want to think about more performant methods of accomplishing this going forward. One option would be to maintain a flag (similar to `_suspend_sync_` that, whenever set, any time we modify `self._data` we write the original unmodified version to a new dict. That way we only backup things that were modified, and we can discard this in an `else` clause to this `try-except` block.